utopia-vscode-extension/src/extension.ts

```diff
@@ -1,14 +1,205 @@
 import * as vscode from 'vscode'
-import { ensureDirectoryExists, waitUntilFSReady } from './browserfs-utils'
-import { fromUtopiaURI, RootDir } from './path-utils'
+import {
+  ensureDirectoryExists,
+  RootDir,
+  initMailbox,
+  isOpenFileMessage,
+  VSCodeInbox,
+  isUpdateDecorationsMessage,
+  DecorationRange,
+  DecorationRangeType,
+  setErrorHandler,
+  FSError,
+  toUtopiaPath,
+  initializeFS,
+  isSelectedElementChanged,
+  BoundsInFile,
+  Bounds,
+  ToVSCodeMessage,
+  parseToVSCodeMessage,
+  sendMessage,
+  editorCursorPositionChanged,
+} from 'utopia-vscode-common'
 import { UtopiaFSExtension } from './utopia-fs'
 
 export async function activate(context: vscode.ExtensionContext): Promise<void> {
   const workspaceRootUri = vscode.workspace.workspaceFolders[0].uri
-  const workspaceRootPath = fromUtopiaURI(workspaceRootUri)
-  await waitUntilFSReady()
+  const projectID = workspaceRootUri.scheme
+  useFileSystemProviderErrors(projectID)
+  await initializeFS(projectID)
   await ensureDirectoryExists(RootDir)
-  await ensureDirectoryExists(workspaceRootPath)
-  const utopiaFS = new UtopiaFSExtension(workspaceRootPath)
+  const utopiaFS = new UtopiaFSExtension(projectID)
   context.subscriptions.push(utopiaFS)
+  initMessaging(utopiaFS, context, workspaceRootUri)
+}
+
+const selectionDecorationType = vscode.window.createTextEditorDecorationType({
+  borderColor: '#007aff',
+  borderStyle: 'solid',
+  borderWidth: '0 0 0 3px',
+  isWholeLine: true,
+})
+
+const highlightDecorationType = vscode.window.createTextEditorDecorationType({
+  borderColor: '#007aff11',
+  borderStyle: 'solid',
+  borderWidth: '0 0 0 3px',
+  isWholeLine: true,
+})
+
+const allDecorationRangeTypes: Array<DecorationRangeType> = ['highlight', 'selection']
+
+function initMessaging(
+  utopiaFS: UtopiaFSExtension,
+  context: vscode.ExtensionContext,
+  workspaceRootUri: vscode.Uri,
+): void {
+  // State that needs to be stored between messages.
+  let currentDecorations: Array<DecorationRange> = []
+
+  function handleMessage(message: ToVSCodeMessage): void {
+    if (isOpenFileMessage(message)) {
+      openFile(utopiaFS, vscode.Uri.joinPath(workspaceRootUri, message.filePath))
+    } else if (isUpdateDecorationsMessage(message)) {
+      currentDecorations = message.decorations
+      updateDecorations(currentDecorations)
+    } else if (isSelectedElementChanged(message)) {
+      revealRangeIfPossible(message.boundsInFile)
+    } else {
+      console.error(`Unhandled message type ${JSON.stringify(message)}`)
```

Comment:
can you have our usual `const _exhaustiveCheck: never = message` here?

Reply:
Probably can, originally the to and from messages shared a type.

Reply:
Done